{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "prospective-knowing",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "saving-tactics",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_df = pd.read_csv('data/processed/AllEpitopeFeatures_larger.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "affiliated-romantic",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>glyco_probs</th>\n",
       "      <th>crosses_cleavage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.035578</td>\n",
       "      <td>0.002893</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.042296</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.042296</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.041116</td>\n",
       "      <td>0.003363</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.038914</td>\n",
       "      <td>0.003206</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163791</th>\n",
       "      <td>S2</td>\n",
       "      <td>KGCCSCGSCCKFDEDDSEPVLKGVK</td>\n",
       "      <td>1244</td>\n",
       "      <td>25</td>\n",
       "      <td>0.116781</td>\n",
       "      <td>0.009853</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163792</th>\n",
       "      <td>S2</td>\n",
       "      <td>GCCSCGSCCKFDEDDSEPVLKGVKL</td>\n",
       "      <td>1245</td>\n",
       "      <td>25</td>\n",
       "      <td>0.113617</td>\n",
       "      <td>0.009618</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163793</th>\n",
       "      <td>S2</td>\n",
       "      <td>CCSCGSCCKFDEDDSEPVLKGVKLH</td>\n",
       "      <td>1246</td>\n",
       "      <td>25</td>\n",
       "      <td>0.113617</td>\n",
       "      <td>0.009618</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163794</th>\n",
       "      <td>S2</td>\n",
       "      <td>CSCGSCCKFDEDDSEPVLKGVKLHY</td>\n",
       "      <td>1247</td>\n",
       "      <td>25</td>\n",
       "      <td>0.099336</td>\n",
       "      <td>0.008367</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163795</th>\n",
       "      <td>S2</td>\n",
       "      <td>SCGSCCKFDEDDSEPVLKGVKLHYT</td>\n",
       "      <td>1248</td>\n",
       "      <td>25</td>\n",
       "      <td>0.098158</td>\n",
       "      <td>0.008289</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163796 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                    epitope  start_pos  epi_len   entropy  \\\n",
       "0            E                   MYSFVSEE          0        8  0.035578   \n",
       "1            E                   YSFVSEET          1        8  0.042296   \n",
       "2            E                   SFVSEETG          2        8  0.042296   \n",
       "3            E                   FVSEETGT          3        8  0.041116   \n",
       "4            E                   VSEETGTL          4        8  0.038914   \n",
       "...        ...                        ...        ...      ...       ...   \n",
       "163791      S2  KGCCSCGSCCKFDEDDSEPVLKGVK       1244       25  0.116781   \n",
       "163792      S2  GCCSCGSCCKFDEDDSEPVLKGVKL       1245       25  0.113617   \n",
       "163793      S2  CCSCGSCCKFDEDDSEPVLKGVKLH       1246       25  0.113617   \n",
       "163794      S2  CSCGSCCKFDEDDSEPVLKGVKLHY       1247       25  0.099336   \n",
       "163795      S2  SCGSCCKFDEDDSEPVLKGVKLHYT       1248       25  0.098158   \n",
       "\n",
       "        perc_mutated  glyco_probs  crosses_cleavage  \n",
       "0           0.002893          0.0                 0  \n",
       "1           0.003441          0.0                 0  \n",
       "2           0.003441          0.0                 0  \n",
       "3           0.003363          0.0                 0  \n",
       "4           0.003206          0.0                 0  \n",
       "...              ...          ...               ...  \n",
       "163791      0.009853          0.0                 0  \n",
       "163792      0.009618          0.0                 0  \n",
       "163793      0.009618          0.0                 0  \n",
       "163794      0.008367          0.0                 0  \n",
       "163795      0.008289          0.0                 0  \n",
       "\n",
       "[163796 rows x 8 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "rational-machine",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>glyco_probs</th>\n",
       "      <th>crosses_cleavage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>142423</th>\n",
       "      <td>S1</td>\n",
       "      <td>MFVFLVLL</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.157835</td>\n",
       "      <td>0.018689</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142424</th>\n",
       "      <td>S1</td>\n",
       "      <td>FVFLVLLP</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.161367</td>\n",
       "      <td>0.018924</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142425</th>\n",
       "      <td>S1</td>\n",
       "      <td>VFLVLLPL</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.159013</td>\n",
       "      <td>0.018768</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142426</th>\n",
       "      <td>S1</td>\n",
       "      <td>FLVLLPLV</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.151556</td>\n",
       "      <td>0.018768</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142427</th>\n",
       "      <td>S1</td>\n",
       "      <td>LVLLPLVS</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.194193</td>\n",
       "      <td>0.023381</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163227</th>\n",
       "      <td>S1</td>\n",
       "      <td>PRRARSVASQSIIAYTMSLGAENSV</td>\n",
       "      <td>680</td>\n",
       "      <td>25</td>\n",
       "      <td>0.454565</td>\n",
       "      <td>0.057945</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163228</th>\n",
       "      <td>S1</td>\n",
       "      <td>RRARSVASQSIIAYTMSLGAENSVA</td>\n",
       "      <td>681</td>\n",
       "      <td>25</td>\n",
       "      <td>0.220743</td>\n",
       "      <td>0.022834</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163229</th>\n",
       "      <td>S1</td>\n",
       "      <td>RARSVASQSIIAYTMSLGAENSVAY</td>\n",
       "      <td>682</td>\n",
       "      <td>25</td>\n",
       "      <td>0.218389</td>\n",
       "      <td>0.022678</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163230</th>\n",
       "      <td>S1</td>\n",
       "      <td>ARSVASQSIIAYTMSLGAENSVAYS</td>\n",
       "      <td>683</td>\n",
       "      <td>25</td>\n",
       "      <td>0.219566</td>\n",
       "      <td>0.022756</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163231</th>\n",
       "      <td>S1</td>\n",
       "      <td>RSVASQSIIAYTMSLGAENSVAYSN</td>\n",
       "      <td>684</td>\n",
       "      <td>25</td>\n",
       "      <td>0.212026</td>\n",
       "      <td>0.022130</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11645 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                    epitope  start_pos  epi_len   entropy  \\\n",
       "142423      S1                   MFVFLVLL          0        8  0.157835   \n",
       "142424      S1                   FVFLVLLP          1        8  0.161367   \n",
       "142425      S1                   VFLVLLPL          2        8  0.159013   \n",
       "142426      S1                   FLVLLPLV          3        8  0.151556   \n",
       "142427      S1                   LVLLPLVS          4        8  0.194193   \n",
       "...        ...                        ...        ...      ...       ...   \n",
       "163227      S1  PRRARSVASQSIIAYTMSLGAENSV        680       25  0.454565   \n",
       "163228      S1  RRARSVASQSIIAYTMSLGAENSVA        681       25  0.220743   \n",
       "163229      S1  RARSVASQSIIAYTMSLGAENSVAY        682       25  0.218389   \n",
       "163230      S1  ARSVASQSIIAYTMSLGAENSVAYS        683       25  0.219566   \n",
       "163231      S1  RSVASQSIIAYTMSLGAENSVAYSN        684       25  0.212026   \n",
       "\n",
       "        perc_mutated  glyco_probs  crosses_cleavage  \n",
       "142423      0.018689          0.0                 0  \n",
       "142424      0.018924          0.0                 0  \n",
       "142425      0.018768          0.0                 0  \n",
       "142426      0.018768          0.0                 0  \n",
       "142427      0.023381          0.0                 0  \n",
       "...              ...          ...               ...  \n",
       "163227      0.057945          0.0                 1  \n",
       "163228      0.022834          0.0                 1  \n",
       "163229      0.022678          0.0                 1  \n",
       "163230      0.022756          0.0                 1  \n",
       "163231      0.022130          1.0                 1  \n",
       "\n",
       "[11645 rows x 8 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df[my_df.protein=='S1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "civic-attack",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "promotional-mayor",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>glyco_probs</th>\n",
       "      <th>crosses_cleavage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>143123</th>\n",
       "      <td>S2</td>\n",
       "      <td>AENSVAYS</td>\n",
       "      <td>700</td>\n",
       "      <td>8</td>\n",
       "      <td>0.122984</td>\n",
       "      <td>0.014779</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144389</th>\n",
       "      <td>S2</td>\n",
       "      <td>AENSVAYSN</td>\n",
       "      <td>700</td>\n",
       "      <td>9</td>\n",
       "      <td>0.124162</td>\n",
       "      <td>0.014858</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145654</th>\n",
       "      <td>S2</td>\n",
       "      <td>AENSVAYSNN</td>\n",
       "      <td>700</td>\n",
       "      <td>10</td>\n",
       "      <td>0.124162</td>\n",
       "      <td>0.014858</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146918</th>\n",
       "      <td>S2</td>\n",
       "      <td>AENSVAYSNNS</td>\n",
       "      <td>700</td>\n",
       "      <td>11</td>\n",
       "      <td>0.125340</td>\n",
       "      <td>0.014936</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148181</th>\n",
       "      <td>S2</td>\n",
       "      <td>AENSVAYSNNSIA</td>\n",
       "      <td>700</td>\n",
       "      <td>13</td>\n",
       "      <td>0.126517</td>\n",
       "      <td>0.015014</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149442</th>\n",
       "      <td>S2</td>\n",
       "      <td>AENSVAYSNNSIAI</td>\n",
       "      <td>700</td>\n",
       "      <td>14</td>\n",
       "      <td>0.127695</td>\n",
       "      <td>0.015092</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150702</th>\n",
       "      <td>S2</td>\n",
       "      <td>AENSVAYSNNSIAIP</td>\n",
       "      <td>700</td>\n",
       "      <td>15</td>\n",
       "      <td>0.127695</td>\n",
       "      <td>0.015092</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151961</th>\n",
       "      <td>S2</td>\n",
       "      <td>AENSVAYSNNSIAIPT</td>\n",
       "      <td>700</td>\n",
       "      <td>16</td>\n",
       "      <td>0.306343</td>\n",
       "      <td>0.042071</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153219</th>\n",
       "      <td>S2</td>\n",
       "      <td>AENSVAYSNNSIAIPTN</td>\n",
       "      <td>700</td>\n",
       "      <td>17</td>\n",
       "      <td>0.308692</td>\n",
       "      <td>0.042227</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154476</th>\n",
       "      <td>S2</td>\n",
       "      <td>AENSVAYSNNSIAIPTNF</td>\n",
       "      <td>700</td>\n",
       "      <td>18</td>\n",
       "      <td>0.308692</td>\n",
       "      <td>0.042227</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155732</th>\n",
       "      <td>S2</td>\n",
       "      <td>AENSVAYSNNSIAIPTNFT</td>\n",
       "      <td>700</td>\n",
       "      <td>19</td>\n",
       "      <td>0.310121</td>\n",
       "      <td>0.042305</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156987</th>\n",
       "      <td>S2</td>\n",
       "      <td>AENSVAYSNNSIAIPTNFTI</td>\n",
       "      <td>700</td>\n",
       "      <td>20</td>\n",
       "      <td>0.313273</td>\n",
       "      <td>0.042540</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158241</th>\n",
       "      <td>S2</td>\n",
       "      <td>AENSVAYSNNSIAIPTNFTIS</td>\n",
       "      <td>700</td>\n",
       "      <td>21</td>\n",
       "      <td>0.316425</td>\n",
       "      <td>0.042774</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159494</th>\n",
       "      <td>S2</td>\n",
       "      <td>AENSVAYSNNSIAIPTNFTISV</td>\n",
       "      <td>700</td>\n",
       "      <td>22</td>\n",
       "      <td>0.316425</td>\n",
       "      <td>0.042774</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160746</th>\n",
       "      <td>S2</td>\n",
       "      <td>AENSVAYSNNSIAIPTNFTISVT</td>\n",
       "      <td>700</td>\n",
       "      <td>23</td>\n",
       "      <td>0.333157</td>\n",
       "      <td>0.044338</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161997</th>\n",
       "      <td>S2</td>\n",
       "      <td>AENSVAYSNNSIAIPTNFTISVTT</td>\n",
       "      <td>700</td>\n",
       "      <td>24</td>\n",
       "      <td>0.334331</td>\n",
       "      <td>0.044417</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163247</th>\n",
       "      <td>S2</td>\n",
       "      <td>AENSVAYSNNSIAIPTNFTISVTTE</td>\n",
       "      <td>700</td>\n",
       "      <td>25</td>\n",
       "      <td>0.334331</td>\n",
       "      <td>0.044417</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                    epitope  start_pos  epi_len   entropy  \\\n",
       "143123      S2                   AENSVAYS        700        8  0.122984   \n",
       "144389      S2                  AENSVAYSN        700        9  0.124162   \n",
       "145654      S2                 AENSVAYSNN        700       10  0.124162   \n",
       "146918      S2                AENSVAYSNNS        700       11  0.125340   \n",
       "148181      S2              AENSVAYSNNSIA        700       13  0.126517   \n",
       "149442      S2             AENSVAYSNNSIAI        700       14  0.127695   \n",
       "150702      S2            AENSVAYSNNSIAIP        700       15  0.127695   \n",
       "151961      S2           AENSVAYSNNSIAIPT        700       16  0.306343   \n",
       "153219      S2          AENSVAYSNNSIAIPTN        700       17  0.308692   \n",
       "154476      S2         AENSVAYSNNSIAIPTNF        700       18  0.308692   \n",
       "155732      S2        AENSVAYSNNSIAIPTNFT        700       19  0.310121   \n",
       "156987      S2       AENSVAYSNNSIAIPTNFTI        700       20  0.313273   \n",
       "158241      S2      AENSVAYSNNSIAIPTNFTIS        700       21  0.316425   \n",
       "159494      S2     AENSVAYSNNSIAIPTNFTISV        700       22  0.316425   \n",
       "160746      S2    AENSVAYSNNSIAIPTNFTISVT        700       23  0.333157   \n",
       "161997      S2   AENSVAYSNNSIAIPTNFTISVTT        700       24  0.334331   \n",
       "163247      S2  AENSVAYSNNSIAIPTNFTISVTTE        700       25  0.334331   \n",
       "\n",
       "        perc_mutated  glyco_probs  crosses_cleavage  \n",
       "143123      0.014779          0.0                 0  \n",
       "144389      0.014858          1.0                 0  \n",
       "145654      0.014858          1.0                 0  \n",
       "146918      0.014936          1.0                 0  \n",
       "148181      0.015014          1.0                 0  \n",
       "149442      0.015092          1.0                 0  \n",
       "150702      0.015092          1.0                 0  \n",
       "151961      0.042071          1.0                 0  \n",
       "153219      0.042227          1.0                 0  \n",
       "154476      0.042227          1.0                 0  \n",
       "155732      0.042305          1.0                 0  \n",
       "156987      0.042540          1.0                 0  \n",
       "158241      0.042774          1.0                 0  \n",
       "159494      0.042774          1.0                 0  \n",
       "160746      0.044338          1.0                 0  \n",
       "161997      0.044417          1.0                 0  \n",
       "163247      0.044417          1.0                 0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# my_df[my_df.protein=='S2']\n",
    "# my_df[my_df.start_pos == 700]\n",
    "my_df[np.logical_and(my_df.start_pos == 700, my_df.protein=='S2')]\n",
    "# len(my_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "capable-craps",
   "metadata": {},
   "source": [
    "### South African Mutation (501Y.V2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "monthly-princeton",
   "metadata": {},
   "outputs": [],
   "source": [
    "### South African Mutation, add a column of binary values showing whether there is a mutation there\n",
    "my_df['South_African_mutation'] = pd.Series(np.zeros(len(my_df)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "comparative-mainland",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>glyco_probs</th>\n",
       "      <th>crosses_cleavage</th>\n",
       "      <th>South_African_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.035578</td>\n",
       "      <td>0.002893</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.042296</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.042296</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.041116</td>\n",
       "      <td>0.003363</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.038914</td>\n",
       "      <td>0.003206</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163791</th>\n",
       "      <td>S2</td>\n",
       "      <td>KGCCSCGSCCKFDEDDSEPVLKGVK</td>\n",
       "      <td>1244</td>\n",
       "      <td>25</td>\n",
       "      <td>0.116781</td>\n",
       "      <td>0.009853</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163792</th>\n",
       "      <td>S2</td>\n",
       "      <td>GCCSCGSCCKFDEDDSEPVLKGVKL</td>\n",
       "      <td>1245</td>\n",
       "      <td>25</td>\n",
       "      <td>0.113617</td>\n",
       "      <td>0.009618</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163793</th>\n",
       "      <td>S2</td>\n",
       "      <td>CCSCGSCCKFDEDDSEPVLKGVKLH</td>\n",
       "      <td>1246</td>\n",
       "      <td>25</td>\n",
       "      <td>0.113617</td>\n",
       "      <td>0.009618</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163794</th>\n",
       "      <td>S2</td>\n",
       "      <td>CSCGSCCKFDEDDSEPVLKGVKLHY</td>\n",
       "      <td>1247</td>\n",
       "      <td>25</td>\n",
       "      <td>0.099336</td>\n",
       "      <td>0.008367</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163795</th>\n",
       "      <td>S2</td>\n",
       "      <td>SCGSCCKFDEDDSEPVLKGVKLHYT</td>\n",
       "      <td>1248</td>\n",
       "      <td>25</td>\n",
       "      <td>0.098158</td>\n",
       "      <td>0.008289</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163796 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                    epitope  start_pos  epi_len   entropy  \\\n",
       "0            E                   MYSFVSEE          0        8  0.035578   \n",
       "1            E                   YSFVSEET          1        8  0.042296   \n",
       "2            E                   SFVSEETG          2        8  0.042296   \n",
       "3            E                   FVSEETGT          3        8  0.041116   \n",
       "4            E                   VSEETGTL          4        8  0.038914   \n",
       "...        ...                        ...        ...      ...       ...   \n",
       "163791      S2  KGCCSCGSCCKFDEDDSEPVLKGVK       1244       25  0.116781   \n",
       "163792      S2  GCCSCGSCCKFDEDDSEPVLKGVKL       1245       25  0.113617   \n",
       "163793      S2  CCSCGSCCKFDEDDSEPVLKGVKLH       1246       25  0.113617   \n",
       "163794      S2  CSCGSCCKFDEDDSEPVLKGVKLHY       1247       25  0.099336   \n",
       "163795      S2  SCGSCCKFDEDDSEPVLKGVKLHYT       1248       25  0.098158   \n",
       "\n",
       "        perc_mutated  glyco_probs  crosses_cleavage  South_African_mutation  \n",
       "0           0.002893          0.0                 0                     0.0  \n",
       "1           0.003441          0.0                 0                     0.0  \n",
       "2           0.003441          0.0                 0                     0.0  \n",
       "3           0.003363          0.0                 0                     0.0  \n",
       "4           0.003206          0.0                 0                     0.0  \n",
       "...              ...          ...               ...                     ...  \n",
       "163791      0.009853          0.0                 0                     0.0  \n",
       "163792      0.009618          0.0                 0                     0.0  \n",
       "163793      0.009618          0.0                 0                     0.0  \n",
       "163794      0.008367          0.0                 0                     0.0  \n",
       "163795      0.008289          0.0                 0                     0.0  \n",
       "\n",
       "[163796 rows x 9 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "thorough-lesson",
   "metadata": {},
   "outputs": [],
   "source": [
    "south_african_indeces = my_df[np.logical_or(my_df.protein=='S1', my_df.protein=='S2')].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "thousand-transcript",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Int64Index([142423, 142424, 142425, 142426, 142427, 142428, 142429, 142430,\n",
       "            142431, 142432,\n",
       "            ...\n",
       "            163786, 163787, 163788, 163789, 163790, 163791, 163792, 163793,\n",
       "            163794, 163795],\n",
       "           dtype='int64', length=21373)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "south_african_indeces"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "considered-sampling",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "antique-sherman",
   "metadata": {},
   "outputs": [],
   "source": [
    "def south_africa(mutation_point):\n",
    "    '''\n",
    "    Inputs:\n",
    "        - mutation point: the number of the nucleotide in the S protein that is mutated\n",
    "    Outputs:\n",
    "        nothing\n",
    "        But, it changes the my_df dataframe\n",
    "    '''\n",
    "    for my_index in south_african_indeces:\n",
    "        temp_start = my_df.loc[my_index, 'start_pos']\n",
    "        temp_len = my_df.loc[my_index, 'epi_len']\n",
    "        if temp_start <= mutation_point and temp_start+temp_len > mutation_point:\n",
    "            my_df.loc[my_index, 'South_African_mutation'] = 1\n",
    "#             print(my_df.loc[my_index,:])\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "alternate-resident",
   "metadata": {},
   "outputs": [],
   "source": [
    "### From this paper: https://www.medrxiv.org/content/10.1101/2020.12.21.20248640v1.full.pdf\n",
    "### Last figure graph B, remember everything here is 0-indexed, so 701 becomes 700\n",
    "### A701V \n",
    "south_africa(700)\n",
    "### D215G\n",
    "south_africa(214)\n",
    "### D80A\n",
    "south_africa(79)\n",
    "### E484K\n",
    "south_africa(483)\n",
    "### K417N\n",
    "south_africa(416)\n",
    "### L18F\n",
    "south_africa(17)\n",
    "### N501Y\n",
    "south_africa(500)\n",
    "### R246I\n",
    "south_africa(245)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "extra-latin",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>glyco_probs</th>\n",
       "      <th>crosses_cleavage</th>\n",
       "      <th>South_African_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.035578</td>\n",
       "      <td>0.002893</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.042296</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.042296</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.041116</td>\n",
       "      <td>0.003363</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.038914</td>\n",
       "      <td>0.003206</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163791</th>\n",
       "      <td>S2</td>\n",
       "      <td>KGCCSCGSCCKFDEDDSEPVLKGVK</td>\n",
       "      <td>1244</td>\n",
       "      <td>25</td>\n",
       "      <td>0.116781</td>\n",
       "      <td>0.009853</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163792</th>\n",
       "      <td>S2</td>\n",
       "      <td>GCCSCGSCCKFDEDDSEPVLKGVKL</td>\n",
       "      <td>1245</td>\n",
       "      <td>25</td>\n",
       "      <td>0.113617</td>\n",
       "      <td>0.009618</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163793</th>\n",
       "      <td>S2</td>\n",
       "      <td>CCSCGSCCKFDEDDSEPVLKGVKLH</td>\n",
       "      <td>1246</td>\n",
       "      <td>25</td>\n",
       "      <td>0.113617</td>\n",
       "      <td>0.009618</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163794</th>\n",
       "      <td>S2</td>\n",
       "      <td>CSCGSCCKFDEDDSEPVLKGVKLHY</td>\n",
       "      <td>1247</td>\n",
       "      <td>25</td>\n",
       "      <td>0.099336</td>\n",
       "      <td>0.008367</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163795</th>\n",
       "      <td>S2</td>\n",
       "      <td>SCGSCCKFDEDDSEPVLKGVKLHYT</td>\n",
       "      <td>1248</td>\n",
       "      <td>25</td>\n",
       "      <td>0.098158</td>\n",
       "      <td>0.008289</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163796 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                    epitope  start_pos  epi_len   entropy  \\\n",
       "0            E                   MYSFVSEE          0        8  0.035578   \n",
       "1            E                   YSFVSEET          1        8  0.042296   \n",
       "2            E                   SFVSEETG          2        8  0.042296   \n",
       "3            E                   FVSEETGT          3        8  0.041116   \n",
       "4            E                   VSEETGTL          4        8  0.038914   \n",
       "...        ...                        ...        ...      ...       ...   \n",
       "163791      S2  KGCCSCGSCCKFDEDDSEPVLKGVK       1244       25  0.116781   \n",
       "163792      S2  GCCSCGSCCKFDEDDSEPVLKGVKL       1245       25  0.113617   \n",
       "163793      S2  CCSCGSCCKFDEDDSEPVLKGVKLH       1246       25  0.113617   \n",
       "163794      S2  CSCGSCCKFDEDDSEPVLKGVKLHY       1247       25  0.099336   \n",
       "163795      S2  SCGSCCKFDEDDSEPVLKGVKLHYT       1248       25  0.098158   \n",
       "\n",
       "        perc_mutated  glyco_probs  crosses_cleavage  South_African_mutation  \n",
       "0           0.002893          0.0                 0                     0.0  \n",
       "1           0.003441          0.0                 0                     0.0  \n",
       "2           0.003441          0.0                 0                     0.0  \n",
       "3           0.003363          0.0                 0                     0.0  \n",
       "4           0.003206          0.0                 0                     0.0  \n",
       "...              ...          ...               ...                     ...  \n",
       "163791      0.009853          0.0                 0                     0.0  \n",
       "163792      0.009618          0.0                 0                     0.0  \n",
       "163793      0.009618          0.0                 0                     0.0  \n",
       "163794      0.008367          0.0                 0                     0.0  \n",
       "163795      0.008289          0.0                 0                     0.0  \n",
       "\n",
       "[163796 rows x 9 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "my_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "vital-virginia",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0    161580\n",
       "1.0      2216\n",
       "Name: South_African_mutation, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df.South_African_mutation.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "tamil-slovakia",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Uncomment to save the file!!!\n",
    "# my_df.to_csv('data/processed/AllEpitopeFeatures_with_South_African_mutation_larger.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "funky-siemens",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "authorized-belarus",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "thrown-supplement",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "synthetic-blood",
   "metadata": {},
   "source": [
    "### UK mutation (B.1.1.7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "allied-interim",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_df = pd.read_csv('data/processed/AllEpitopeFeatures_with_South_African_mutation_larger.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "republican-mention",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_df;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "vanilla-dispatch",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_df['UK_mutation'] = pd.Series(np.zeros(len(my_df)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "eligible-kitty",
   "metadata": {},
   "outputs": [],
   "source": [
    "# my_df.UK_mutation.value_counts()\n",
    "my_df.protein.unique();\n",
    "my_df[my_df.protein=='ORF1a'];"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "offshore-webmaster",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_df[np.logical_and(my_df.protein=='ORF1a', my_df.start_pos==1000)];"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "wicked-classification",
   "metadata": {},
   "outputs": [],
   "source": [
    "def uk_mutation(mutation_point, my_protein):\n",
    "    '''\n",
    "    Inputs:\n",
    "        - mutation point: the number of the nucleotide in the S protein that is mutated\n",
    "        - protein: the protein, ex. S (S1 and S2), ORF1a (there is no need to say ORF1ab as \n",
    "        there are no mutations past the cutoff between ORF1a and ORF1b) etc.\n",
    "    Outputs:\n",
    "        nothing\n",
    "        But, it changes the my_df dataframe\n",
    "    '''\n",
    "    if my_protein == 'S':\n",
    "        indeces = my_df[np.logical_or(my_df.protein=='S1', my_df.protein=='S2')].index\n",
    "    else:\n",
    "        indeces = my_df[my_df.protein==my_protein].index\n",
    "        \n",
    "    for my_index in indeces:\n",
    "        temp_start = my_df.loc[my_index, 'start_pos']\n",
    "        temp_len = my_df.loc[my_index, 'epi_len']\n",
    "        if temp_start <= mutation_point and temp_start+temp_len > mutation_point:\n",
    "            my_df.loc[my_index, 'UK_mutation'] = 1\n",
    "#             print(my_index)\n",
    "#             print(my_df.loc[my_index,:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "associate-fundamentals",
   "metadata": {},
   "outputs": [],
   "source": [
    "### From this paper: https://virological.org/t/preliminary-genomic-characterisation-of-an-emergent-sars-cov-2-lineage-in-the-uk-defined-by-a-novel-set-of-spike-mutations/563\n",
    "### 0-indexed!!!, also instead of ORF1ab I use ORF1a, because all the mutations are in the beginning of \n",
    "### the ORF1ab, so in ORF1a\n",
    "\n",
    "### ORF1ab -> ORF1a\n",
    "### T1001I\n",
    "uk_mutation(1000, 'ORF1a')\n",
    "### A1708D\n",
    "uk_mutation(1707, 'ORF1a')\n",
    "### I2230T\n",
    "uk_mutation(2229, 'ORF1a')\n",
    "### SGF 3675-3677 deletion\n",
    "uk_mutation(3674, 'ORF1a')\n",
    "uk_mutation(3675, 'ORF1a')\n",
    "uk_mutation(3676, 'ORF1a')\n",
    "\n",
    "### S (Spike protein)\n",
    "### HV 69-70 deletion\n",
    "uk_mutation(68, 'S')\n",
    "uk_mutation(69, 'S')\n",
    "### Y144 deletion\n",
    "uk_mutation(143, 'S')\n",
    "### N501Y\n",
    "uk_mutation(500, 'S')\n",
    "### A570D\n",
    "uk_mutation(569, 'S')\n",
    "### P681H\n",
    "uk_mutation(680, 'S')\n",
    "### T716I\n",
    "uk_mutation(715, 'S')\n",
    "### S982A\n",
    "uk_mutation(981, 'S')\n",
    "### D1118H\n",
    "uk_mutation(1117, 'S')\n",
    "\n",
    "### ORF8\n",
    "### Q27stop -> Basically get rid of all of ORF8 !!!!!\n",
    "\n",
    "# ### R52I\n",
    "# uk_mutation(51, 'ORF8')\n",
    "# ### Y73C\n",
    "# uk_mutation(72, 'ORF8')\n",
    "\n",
    "### N\n",
    "### D3L\n",
    "uk_mutation(2, 'N')\n",
    "### S235F\n",
    "uk_mutation(234, 'N')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "afraid-washer",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0    159989\n",
       "1.0      3807\n",
       "Name: UK_mutation, dtype: int64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df.UK_mutation.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "intimate-passion",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Int64Index([139253, 139254, 139255, 139256, 139257, 139258, 139259, 139260,\n",
       "            139261, 139262,\n",
       "            ...\n",
       "            141032, 141033, 141034, 141035, 141036, 141037, 141038, 141039,\n",
       "            141040, 141041],\n",
       "           dtype='int64', length=1789)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Take care of all of ORF8\n",
    "indeces_orf8 = my_df[my_df.protein=='ORF8'].index\n",
    "indeces_orf8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "typical-chemistry",
   "metadata": {},
   "outputs": [],
   "source": [
    "for my_index in indeces_orf8:\n",
    "    my_df.loc[my_index, 'UK_mutation'] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "saving-accounting",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0    158200\n",
       "1.0      5596\n",
       "Name: UK_mutation, dtype: int64"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df.UK_mutation.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "bearing-uniform",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Uncomment to save the file\n",
    "# my_df.to_csv('data/processed/AllEpitopeFeatures_with_South_African_and_UK_mutation_larger.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "explicit-lawsuit",
   "metadata": {},
   "source": [
    "### Indian Mutation (B.1.617)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "handled-collectible",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Use paper: https://www.biorxiv.org/content/10.1101/2021.04.23.441101v1.full"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "foster-opening",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_df = pd.read_csv('data/processed/AllEpitopeFeatures_with_South_African_and_UK_mutation_larger.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "tender-pillow",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>glyco_probs</th>\n",
       "      <th>crosses_cleavage</th>\n",
       "      <th>South_African_mutation</th>\n",
       "      <th>UK_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.035578</td>\n",
       "      <td>0.002893</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.042296</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.042296</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.041116</td>\n",
       "      <td>0.003363</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.038914</td>\n",
       "      <td>0.003206</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163791</th>\n",
       "      <td>S2</td>\n",
       "      <td>KGCCSCGSCCKFDEDDSEPVLKGVK</td>\n",
       "      <td>1244</td>\n",
       "      <td>25</td>\n",
       "      <td>0.116781</td>\n",
       "      <td>0.009853</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163792</th>\n",
       "      <td>S2</td>\n",
       "      <td>GCCSCGSCCKFDEDDSEPVLKGVKL</td>\n",
       "      <td>1245</td>\n",
       "      <td>25</td>\n",
       "      <td>0.113617</td>\n",
       "      <td>0.009618</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163793</th>\n",
       "      <td>S2</td>\n",
       "      <td>CCSCGSCCKFDEDDSEPVLKGVKLH</td>\n",
       "      <td>1246</td>\n",
       "      <td>25</td>\n",
       "      <td>0.113617</td>\n",
       "      <td>0.009618</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163794</th>\n",
       "      <td>S2</td>\n",
       "      <td>CSCGSCCKFDEDDSEPVLKGVKLHY</td>\n",
       "      <td>1247</td>\n",
       "      <td>25</td>\n",
       "      <td>0.099336</td>\n",
       "      <td>0.008367</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163795</th>\n",
       "      <td>S2</td>\n",
       "      <td>SCGSCCKFDEDDSEPVLKGVKLHYT</td>\n",
       "      <td>1248</td>\n",
       "      <td>25</td>\n",
       "      <td>0.098158</td>\n",
       "      <td>0.008289</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163796 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                    epitope  start_pos  epi_len   entropy  \\\n",
       "0            E                   MYSFVSEE          0        8  0.035578   \n",
       "1            E                   YSFVSEET          1        8  0.042296   \n",
       "2            E                   SFVSEETG          2        8  0.042296   \n",
       "3            E                   FVSEETGT          3        8  0.041116   \n",
       "4            E                   VSEETGTL          4        8  0.038914   \n",
       "...        ...                        ...        ...      ...       ...   \n",
       "163791      S2  KGCCSCGSCCKFDEDDSEPVLKGVK       1244       25  0.116781   \n",
       "163792      S2  GCCSCGSCCKFDEDDSEPVLKGVKL       1245       25  0.113617   \n",
       "163793      S2  CCSCGSCCKFDEDDSEPVLKGVKLH       1246       25  0.113617   \n",
       "163794      S2  CSCGSCCKFDEDDSEPVLKGVKLHY       1247       25  0.099336   \n",
       "163795      S2  SCGSCCKFDEDDSEPVLKGVKLHYT       1248       25  0.098158   \n",
       "\n",
       "        perc_mutated  glyco_probs  crosses_cleavage  South_African_mutation  \\\n",
       "0           0.002893          0.0                 0                     0.0   \n",
       "1           0.003441          0.0                 0                     0.0   \n",
       "2           0.003441          0.0                 0                     0.0   \n",
       "3           0.003363          0.0                 0                     0.0   \n",
       "4           0.003206          0.0                 0                     0.0   \n",
       "...              ...          ...               ...                     ...   \n",
       "163791      0.009853          0.0                 0                     0.0   \n",
       "163792      0.009618          0.0                 0                     0.0   \n",
       "163793      0.009618          0.0                 0                     0.0   \n",
       "163794      0.008367          0.0                 0                     0.0   \n",
       "163795      0.008289          0.0                 0                     0.0   \n",
       "\n",
       "        UK_mutation  \n",
       "0               0.0  \n",
       "1               0.0  \n",
       "2               0.0  \n",
       "3               0.0  \n",
       "4               0.0  \n",
       "...             ...  \n",
       "163791          0.0  \n",
       "163792          0.0  \n",
       "163793          0.0  \n",
       "163794          0.0  \n",
       "163795          0.0  \n",
       "\n",
       "[163796 rows x 10 columns]"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "olive-skating",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_df['Indian_mutation'] = pd.Series(np.zeros(len(my_df)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "emerging-ordinary",
   "metadata": {},
   "outputs": [],
   "source": [
    "def indian_mutation(mutation_point, my_protein):\n",
    "    '''\n",
    "    Inputs:\n",
    "        - mutation point: the number of the nucleotide in the S protein that is mutated\n",
    "        - protein: the protein, ex. S (S1 and S2), ORF1a, ORF1b etc.\n",
    "    Outputs:\n",
    "        nothing\n",
    "        But, it changes the my_df dataframe\n",
    "    '''\n",
    "#     if my_protein == 'S':\n",
    "#         indeces = my_df[np.logical_or(my_df.protein=='S1', my_df.protein=='S2')].index\n",
    "#     elif my_protein == '':\n",
    "#         indeces = my_df[my_df.protein==my_protein].index\n",
    "\n",
    "    indeces = my_df[my_df.protein==my_protein].index\n",
    "    \n",
    "    anything = False\n",
    "        \n",
    "    for my_index in indeces:\n",
    "        temp_start = my_df.loc[my_index, 'start_pos']\n",
    "        temp_len = my_df.loc[my_index, 'epi_len']\n",
    "        if temp_start <= mutation_point and temp_start+temp_len > mutation_point:\n",
    "            my_df.loc[my_index, 'Indian_mutation'] = 1\n",
    "            anything = True\n",
    "            \n",
    "    if not anything:\n",
    "        print('Problem with ', mutation_point, ' in protein ', my_protein)\n",
    "#             print(my_index)\n",
    "#             print(my_df.loc[my_index,:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "oriental-pitch",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Mutation points for the Indian Mutation (B.1.617)\n",
    "### From the paper: https://www.biorxiv.org/content/10.1101/2021.04.23.441101v1.full\n",
    "### For all the proteins the mutation point is -1 from what is written, because of zero indexing\n",
    "### Exception: ORF1b, where the mutation point is -3 - 4399, because the numbering continues over from ORF1a\n",
    "### Note that we need not worry in the same way about S2, as the numbering simply continues on there.\n",
    "\n",
    "### ORF1a\n",
    "### T 1567 I \n",
    "indian_mutation(1567 - 1, 'ORF1a')\n",
    "### T 3646 A\n",
    "indian_mutation(3646 - 1, 'ORF1a')\n",
    "\n",
    "### ORF1b \n",
    "### P 4715 L\n",
    "indian_mutation(4715 - 3 - 4399, 'ORF1b')\n",
    "### G 5530 I\n",
    "indian_mutation(5530 - 3 - 4399, 'ORF1b')\n",
    "### M 5753 I\n",
    "indian_mutation(5753 - 3 - 4399, 'ORF1b')\n",
    "### K 6711 R\n",
    "indian_mutation(6711 - 3 - 4399, 'ORF1b')\n",
    "### S 6713 A\n",
    "indian_mutation(6713 - 3 - 4399, 'ORF1b')\n",
    "\n",
    "### S protein\n",
    "### S1 domain\n",
    "### T 95 I\n",
    "indian_mutation(95 - 1, 'S1')\n",
    "### G 142 D\n",
    "indian_mutation(142 - 1, 'S1')\n",
    "### E 154 K\n",
    "indian_mutation(154 - 1, 'S1')\n",
    "### L 452 R\n",
    "indian_mutation(452 - 1, 'S1')\n",
    "### E 484 Q\n",
    "indian_mutation(484 - 1, 'S1')\n",
    "### D 614 G\n",
    "indian_mutation(614 - 1, 'S1')\n",
    "### P 681 R\n",
    "indian_mutation(681 - 1, 'S1')\n",
    "### S2 domain\n",
    "### Q 1071 H\n",
    "indian_mutation(1071 - 1, 'S2')\n",
    "\n",
    "### ORF3a\n",
    "### S 26 L\n",
    "indian_mutation(26 - 1, 'ORF3a')\n",
    "\n",
    "### M\n",
    "### I 82 S\n",
    "indian_mutation(82 - 1, 'M')\n",
    "\n",
    "### ORF6\n",
    "### D 61 (the last amino acid)\n",
    "indian_mutation(61 - 1, 'ORF6')\n",
    "### 62 (anything after that)\n",
    "\n",
    "### ORF7a\n",
    "### N 43 Y\n",
    "indian_mutation(43 - 1, 'ORF7a')\n",
    "### V 82 A\n",
    "indian_mutation(82 - 1, 'ORF7a')\n",
    "\n",
    "### N\n",
    "### R 203 M\n",
    "indian_mutation(203 - 1, 'N')\n",
    "### D 377 Y\n",
    "indian_mutation(377 - 1, 'N')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "christian-dinner",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0    158136\n",
       "1.0      5660\n",
       "Name: Indian_mutation, dtype: int64"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df.Indian_mutation.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "starting-sleeve",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Code to save the data\n",
    "# my_df.to_csv('data/processed/AllEpitopeFeatures_with_South_African_and_UK_and_Indian_mutation_larger.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "composite-verification",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "republican-function",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([   0,    1,    2, ..., 4390, 4391, 4392])"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df[my_df.protein == 'ORF1a'].start_pos.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "offensive-enzyme",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>glyco_probs</th>\n",
       "      <th>crosses_cleavage</th>\n",
       "      <th>South_African_mutation</th>\n",
       "      <th>UK_mutation</th>\n",
       "      <th>Indian_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16138</th>\n",
       "      <td>ORF1a</td>\n",
       "      <td>SADAQSFL</td>\n",
       "      <td>4392</td>\n",
       "      <td>8</td>\n",
       "      <td>0.018261</td>\n",
       "      <td>0.001408</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      protein   epitope  start_pos  epi_len   entropy  perc_mutated  \\\n",
       "16138   ORF1a  SADAQSFL       4392        8  0.018261      0.001408   \n",
       "\n",
       "       glyco_probs  crosses_cleavage  South_African_mutation  UK_mutation  \\\n",
       "16138          0.0                 0                     0.0          0.0   \n",
       "\n",
       "       Indian_mutation  \n",
       "16138              0.0  "
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df[np.logical_and(my_df.protein == 'ORF1a', my_df.start_pos == 4392)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "visible-privacy",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([   0,    1,    2, ..., 2685, 2686, 2687])"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df[my_df.protein == 'ORF1b'].start_pos.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "cellular-blond",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "684"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df[my_df.protein == 'S1'].start_pos.unique()[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "loaded-darwin",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "685"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df[my_df.protein == 'S2'].start_pos.unique()[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "interested-shopper",
   "metadata": {},
   "outputs": [],
   "source": [
    "### ORF1b is after ORF1a, but starts with 0\n",
    "### S2 is after S1, but starts at 685, while S1 ends with 684"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "nuclear-hydrogen",
   "metadata": {},
   "outputs": [],
   "source": [
    "# my_df[np.logical_and(my_df.protein == 'ORF1b', my_df.start_pos == 4712 - 4399)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "developed-gardening",
   "metadata": {},
   "outputs": [],
   "source": [
    "# my_df[np.logical_and(my_df.protein == 'ORF1b', my_df.start_pos == 5529 - 4392)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "pregnant-amendment",
   "metadata": {},
   "outputs": [],
   "source": [
    "# my_df[np.logical_and(my_df.protein == 'N', my_df.start_pos == 376)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "sonic-haven",
   "metadata": {},
   "outputs": [],
   "source": [
    "# my_df.protein.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "nuclear-black",
   "metadata": {},
   "outputs": [],
   "source": [
    "# my_df[my_df.protein == 'ORF6']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "worst-brighton",
   "metadata": {},
   "outputs": [],
   "source": [
    "# my_df[np.logical_and(my_df.protein == 'S1', my_df.start_pos == 245)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "maritime-figure",
   "metadata": {},
   "outputs": [],
   "source": [
    "# my_df[np.logical_and(my_df.protein == 'N', my_df.start_pos == 2)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "id": "conservative-statistics",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>glyco_probs</th>\n",
       "      <th>crosses_cleavage</th>\n",
       "      <th>South_African_mutation</th>\n",
       "      <th>UK_mutation</th>\n",
       "      <th>Indian_mutation</th>\n",
       "      <th>Brazilian_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>88589</th>\n",
       "      <td>ORF1b</td>\n",
       "      <td>SPFELEDF</td>\n",
       "      <td>2311</td>\n",
       "      <td>8</td>\n",
       "      <td>0.042578</td>\n",
       "      <td>0.003363</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91277</th>\n",
       "      <td>ORF1b</td>\n",
       "      <td>SPFELEDFI</td>\n",
       "      <td>2311</td>\n",
       "      <td>9</td>\n",
       "      <td>0.046116</td>\n",
       "      <td>0.003597</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93964</th>\n",
       "      <td>ORF1b</td>\n",
       "      <td>SPFELEDFIP</td>\n",
       "      <td>2311</td>\n",
       "      <td>10</td>\n",
       "      <td>0.053588</td>\n",
       "      <td>0.004145</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96650</th>\n",
       "      <td>ORF1b</td>\n",
       "      <td>SPFELEDFIPM</td>\n",
       "      <td>2311</td>\n",
       "      <td>11</td>\n",
       "      <td>0.055946</td>\n",
       "      <td>0.004301</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99335</th>\n",
       "      <td>ORF1b</td>\n",
       "      <td>SPFELEDFIPMDS</td>\n",
       "      <td>2311</td>\n",
       "      <td>13</td>\n",
       "      <td>0.066672</td>\n",
       "      <td>0.005083</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102018</th>\n",
       "      <td>ORF1b</td>\n",
       "      <td>SPFELEDFIPMDST</td>\n",
       "      <td>2311</td>\n",
       "      <td>14</td>\n",
       "      <td>0.066672</td>\n",
       "      <td>0.005083</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104700</th>\n",
       "      <td>ORF1b</td>\n",
       "      <td>SPFELEDFIPMDSTV</td>\n",
       "      <td>2311</td>\n",
       "      <td>15</td>\n",
       "      <td>0.066672</td>\n",
       "      <td>0.005083</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107381</th>\n",
       "      <td>ORF1b</td>\n",
       "      <td>SPFELEDFIPMDSTVK</td>\n",
       "      <td>2311</td>\n",
       "      <td>16</td>\n",
       "      <td>0.067851</td>\n",
       "      <td>0.005161</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110061</th>\n",
       "      <td>ORF1b</td>\n",
       "      <td>SPFELEDFIPMDSTVKN</td>\n",
       "      <td>2311</td>\n",
       "      <td>17</td>\n",
       "      <td>0.069030</td>\n",
       "      <td>0.005239</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112740</th>\n",
       "      <td>ORF1b</td>\n",
       "      <td>SPFELEDFIPMDSTVKNY</td>\n",
       "      <td>2311</td>\n",
       "      <td>18</td>\n",
       "      <td>0.071232</td>\n",
       "      <td>0.005396</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115418</th>\n",
       "      <td>ORF1b</td>\n",
       "      <td>SPFELEDFIPMDSTVKNYF</td>\n",
       "      <td>2311</td>\n",
       "      <td>19</td>\n",
       "      <td>0.071232</td>\n",
       "      <td>0.005396</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118095</th>\n",
       "      <td>ORF1b</td>\n",
       "      <td>SPFELEDFIPMDSTVKNYFI</td>\n",
       "      <td>2311</td>\n",
       "      <td>20</td>\n",
       "      <td>0.073433</td>\n",
       "      <td>0.005552</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120771</th>\n",
       "      <td>ORF1b</td>\n",
       "      <td>SPFELEDFIPMDSTVKNYFIT</td>\n",
       "      <td>2311</td>\n",
       "      <td>21</td>\n",
       "      <td>0.077836</td>\n",
       "      <td>0.005865</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123446</th>\n",
       "      <td>ORF1b</td>\n",
       "      <td>SPFELEDFIPMDSTVKNYFITD</td>\n",
       "      <td>2311</td>\n",
       "      <td>22</td>\n",
       "      <td>0.083105</td>\n",
       "      <td>0.006256</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126120</th>\n",
       "      <td>ORF1b</td>\n",
       "      <td>SPFELEDFIPMDSTVKNYFITDA</td>\n",
       "      <td>2311</td>\n",
       "      <td>23</td>\n",
       "      <td>0.086485</td>\n",
       "      <td>0.006490</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128793</th>\n",
       "      <td>ORF1b</td>\n",
       "      <td>SPFELEDFIPMDSTVKNYFITDAQ</td>\n",
       "      <td>2311</td>\n",
       "      <td>24</td>\n",
       "      <td>0.086485</td>\n",
       "      <td>0.006490</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131465</th>\n",
       "      <td>ORF1b</td>\n",
       "      <td>SPFELEDFIPMDSTVKNYFITDAQT</td>\n",
       "      <td>2311</td>\n",
       "      <td>25</td>\n",
       "      <td>0.090574</td>\n",
       "      <td>0.006803</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                    epitope  start_pos  epi_len   entropy  \\\n",
       "88589    ORF1b                   SPFELEDF       2311        8  0.042578   \n",
       "91277    ORF1b                  SPFELEDFI       2311        9  0.046116   \n",
       "93964    ORF1b                 SPFELEDFIP       2311       10  0.053588   \n",
       "96650    ORF1b                SPFELEDFIPM       2311       11  0.055946   \n",
       "99335    ORF1b              SPFELEDFIPMDS       2311       13  0.066672   \n",
       "102018   ORF1b             SPFELEDFIPMDST       2311       14  0.066672   \n",
       "104700   ORF1b            SPFELEDFIPMDSTV       2311       15  0.066672   \n",
       "107381   ORF1b           SPFELEDFIPMDSTVK       2311       16  0.067851   \n",
       "110061   ORF1b          SPFELEDFIPMDSTVKN       2311       17  0.069030   \n",
       "112740   ORF1b         SPFELEDFIPMDSTVKNY       2311       18  0.071232   \n",
       "115418   ORF1b        SPFELEDFIPMDSTVKNYF       2311       19  0.071232   \n",
       "118095   ORF1b       SPFELEDFIPMDSTVKNYFI       2311       20  0.073433   \n",
       "120771   ORF1b      SPFELEDFIPMDSTVKNYFIT       2311       21  0.077836   \n",
       "123446   ORF1b     SPFELEDFIPMDSTVKNYFITD       2311       22  0.083105   \n",
       "126120   ORF1b    SPFELEDFIPMDSTVKNYFITDA       2311       23  0.086485   \n",
       "128793   ORF1b   SPFELEDFIPMDSTVKNYFITDAQ       2311       24  0.086485   \n",
       "131465   ORF1b  SPFELEDFIPMDSTVKNYFITDAQT       2311       25  0.090574   \n",
       "\n",
       "        perc_mutated  glyco_probs  crosses_cleavage  South_African_mutation  \\\n",
       "88589       0.003363          0.0                 0                     0.0   \n",
       "91277       0.003597          0.0                 0                     0.0   \n",
       "93964       0.004145          0.0                 0                     0.0   \n",
       "96650       0.004301          0.0                 0                     0.0   \n",
       "99335       0.005083          0.0                 0                     0.0   \n",
       "102018      0.005083          0.0                 0                     0.0   \n",
       "104700      0.005083          0.0                 0                     0.0   \n",
       "107381      0.005161          0.0                 0                     0.0   \n",
       "110061      0.005239          0.0                 0                     0.0   \n",
       "112740      0.005396          0.0                 0                     0.0   \n",
       "115418      0.005396          0.0                 0                     0.0   \n",
       "118095      0.005552          0.0                 0                     0.0   \n",
       "120771      0.005865          0.0                 0                     0.0   \n",
       "123446      0.006256          0.0                 0                     0.0   \n",
       "126120      0.006490          0.0                 0                     0.0   \n",
       "128793      0.006490          0.0                 0                     0.0   \n",
       "131465      0.006803          0.0                 0                     0.0   \n",
       "\n",
       "        UK_mutation  Indian_mutation  Brazilian_mutation  \n",
       "88589           0.0              1.0                 0.0  \n",
       "91277           0.0              1.0                 0.0  \n",
       "93964           0.0              1.0                 0.0  \n",
       "96650           0.0              1.0                 0.0  \n",
       "99335           0.0              1.0                 0.0  \n",
       "102018          0.0              1.0                 0.0  \n",
       "104700          0.0              1.0                 0.0  \n",
       "107381          0.0              1.0                 0.0  \n",
       "110061          0.0              1.0                 0.0  \n",
       "112740          0.0              1.0                 0.0  \n",
       "115418          0.0              1.0                 0.0  \n",
       "118095          0.0              1.0                 0.0  \n",
       "120771          0.0              1.0                 0.0  \n",
       "123446          0.0              1.0                 0.0  \n",
       "126120          0.0              1.0                 0.0  \n",
       "128793          0.0              1.0                 0.0  \n",
       "131465          0.0              1.0                 0.0  "
      ]
     },
     "execution_count": 250,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df[np.logical_and(my_df.protein == 'ORF1b', my_df.start_pos == 6713 - 3 - 4399)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "positive-namibia",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "divided-stand",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "conscious-dictionary",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "meaning-sigma",
   "metadata": {},
   "source": [
    "### Brazilian Mutation (P.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "leading-injury",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Use the following paper: https://virological.org/t/genomic-characterisation-of-an-emergent-sars-cov-2-lineage-in-manaus-preliminary-findings/586"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "id": "sublime-technical",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_df = pd.read_csv('data/processed/AllEpitopeFeatures_with_South_African_and_UK_and_Indian_mutation_larger.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "id": "important-lucas",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0    158136\n",
       "1.0      5660\n",
       "Name: Indian_mutation, dtype: int64"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df.Indian_mutation.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "id": "upset-contact",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_df['Brazilian_mutation'] = pd.Series(np.zeros(len(my_df)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "speaking-japanese",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "hybrid-density",
   "metadata": {},
   "outputs": [],
   "source": [
    "def brazilian_mutation(mutation_point, my_protein):\n",
    "    ### exact same function as the indian_mutation !!!\n",
    "    '''\n",
    "    Inputs:\n",
    "        - mutation point: the number of the nucleotide in the S protein that is mutated\n",
    "        - protein: the protein, ex. S (S1 and S2), ORF1a, ORF1b etc.\n",
    "    Outputs:\n",
    "        nothing\n",
    "        But, it changes the my_df dataframe\n",
    "    '''\n",
    "#     if my_protein == 'S':\n",
    "#         indeces = my_df[np.logical_or(my_df.protein=='S1', my_df.protein=='S2')].index\n",
    "#     elif my_protein == '':\n",
    "#         indeces = my_df[my_df.protein==my_protein].index\n",
    "\n",
    "    indeces = my_df[my_df.protein==my_protein].index\n",
    "    \n",
    "    anything = False\n",
    "        \n",
    "    for my_index in indeces:\n",
    "        temp_start = my_df.loc[my_index, 'start_pos']\n",
    "        temp_len = my_df.loc[my_index, 'epi_len']\n",
    "        if temp_start <= mutation_point and temp_start+temp_len > mutation_point:\n",
    "            my_df.loc[my_index, 'Brazilian_mutation'] = 1\n",
    "            anything = True\n",
    "            \n",
    "    if not anything:\n",
    "        print('Problem with ', mutation_point, ' in protein ', my_protein)\n",
    "#             print(my_index)\n",
    "#             print(my_df.loc[my_index,:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "foster-moore",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Mutations from the paper\n",
    "### Paper: https://virological.org/t/genomic-characterisation-of-an-emergent-sars-cov-2-lineage-in-manaus-preliminary-findings/586\n",
    "\n",
    "### For all the proteins the mutation point is -1 from what is written, because of zero indexing\n",
    "### Exception: ORF1b, where the mutation point is -3 - 4399, because the numbering continues over from ORF1a\n",
    "### Note that we need not worry in the same way about S2, as the numbering simply continues on there.\n",
    "\n",
    "\n",
    "\n",
    "### ORF1ab\n",
    "### ORF1a\n",
    "### S 1188 L\n",
    "brazilian_mutation(1188 - 1, 'ORF1a')\n",
    "### K 1795 Q\n",
    "brazilian_mutation(1795 - 1, 'ORF1a')\n",
    "### del 3675\n",
    "brazilian_mutation(3675 - 1, 'ORF1a')\n",
    "### del 3676\n",
    "brazilian_mutation(3676 - 1, 'ORF1a')\n",
    "### del 3677\n",
    "brazilian_mutation(3677 - 1, 'ORF1a')\n",
    "### ORF1b\n",
    "### E 5665 D\n",
    "brazilian_mutation(5665 - 3 - 4399, 'ORF1b')\n",
    "\n",
    "### S (S1 and S2)\n",
    "### S1\n",
    "### L 18 F\n",
    "brazilian_mutation(18 - 1, 'S1')\n",
    "### T 20 N\n",
    "brazilian_mutation(20 - 1, 'S1')\n",
    "### P 26 S\n",
    "brazilian_mutation(26 - 1, 'S1')\n",
    "### D 138 Y\n",
    "brazilian_mutation(138 - 1, 'S1')\n",
    "### R 190 S\n",
    "brazilian_mutation(190 - 1, 'S1')\n",
    "### K 417 T (important!!!)\n",
    "brazilian_mutation(417 - 1, 'S1')\n",
    "### E 484 K (important!!!)\n",
    "brazilian_mutation(484 - 1, 'S1')\n",
    "### N 501 Y (important!!!)\n",
    "brazilian_mutation(501 - 1, 'S1')\n",
    "### H 655 Y\n",
    "brazilian_mutation(655 - 1, 'S1')\n",
    "### S2\n",
    "### T 1027 I\n",
    "brazilian_mutation(1027 - 1, 'S2')\n",
    "\n",
    "### ORF8\n",
    "### E 92 K\n",
    "brazilian_mutation(92 - 1, 'ORF8')\n",
    "### insertion in DNA 28269 - 28273 ???\n",
    "\n",
    "### N\n",
    "### P 80 R\n",
    "brazilian_mutation(80 - 1, 'N')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "id": "cubic-spokesman",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Checking that things work\n",
    "# my_df[np.logical_and(my_df.protein == 'ORF1b', my_df.start_pos == 5665 - 3 - 4399)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "killing-nepal",
   "metadata": {},
   "outputs": [],
   "source": [
    "# my_df[np.logical_and(my_df.protein == 'N', my_df.start_pos == 79)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "published-enterprise",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0    159700\n",
       "1.0      4096\n",
       "Name: Brazilian_mutation, dtype: int64"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df.Brazilian_mutation.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "white-gambling",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>glyco_probs</th>\n",
       "      <th>crosses_cleavage</th>\n",
       "      <th>South_African_mutation</th>\n",
       "      <th>UK_mutation</th>\n",
       "      <th>Indian_mutation</th>\n",
       "      <th>Brazilian_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4592</th>\n",
       "      <td>N</td>\n",
       "      <td>PDDQIGYY</td>\n",
       "      <td>79</td>\n",
       "      <td>8</td>\n",
       "      <td>0.030754</td>\n",
       "      <td>0.002581</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5004</th>\n",
       "      <td>N</td>\n",
       "      <td>PDDQIGYYR</td>\n",
       "      <td>79</td>\n",
       "      <td>9</td>\n",
       "      <td>0.030754</td>\n",
       "      <td>0.002581</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5415</th>\n",
       "      <td>N</td>\n",
       "      <td>PDDQIGYYRR</td>\n",
       "      <td>79</td>\n",
       "      <td>10</td>\n",
       "      <td>0.031933</td>\n",
       "      <td>0.002659</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5825</th>\n",
       "      <td>N</td>\n",
       "      <td>PDDQIGYYRRA</td>\n",
       "      <td>79</td>\n",
       "      <td>11</td>\n",
       "      <td>0.039191</td>\n",
       "      <td>0.003206</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6234</th>\n",
       "      <td>N</td>\n",
       "      <td>PDDQIGYYRRATR</td>\n",
       "      <td>79</td>\n",
       "      <td>13</td>\n",
       "      <td>0.039191</td>\n",
       "      <td>0.003206</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6641</th>\n",
       "      <td>N</td>\n",
       "      <td>PDDQIGYYRRATRR</td>\n",
       "      <td>79</td>\n",
       "      <td>14</td>\n",
       "      <td>0.039191</td>\n",
       "      <td>0.003206</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7047</th>\n",
       "      <td>N</td>\n",
       "      <td>PDDQIGYYRRATRRI</td>\n",
       "      <td>79</td>\n",
       "      <td>15</td>\n",
       "      <td>0.039191</td>\n",
       "      <td>0.003206</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7452</th>\n",
       "      <td>N</td>\n",
       "      <td>PDDQIGYYRRATRRIR</td>\n",
       "      <td>79</td>\n",
       "      <td>16</td>\n",
       "      <td>0.040370</td>\n",
       "      <td>0.003284</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7856</th>\n",
       "      <td>N</td>\n",
       "      <td>PDDQIGYYRRATRRIRG</td>\n",
       "      <td>79</td>\n",
       "      <td>17</td>\n",
       "      <td>0.042728</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8259</th>\n",
       "      <td>N</td>\n",
       "      <td>PDDQIGYYRRATRRIRGG</td>\n",
       "      <td>79</td>\n",
       "      <td>18</td>\n",
       "      <td>0.042728</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8661</th>\n",
       "      <td>N</td>\n",
       "      <td>PDDQIGYYRRATRRIRGGD</td>\n",
       "      <td>79</td>\n",
       "      <td>19</td>\n",
       "      <td>0.042728</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9062</th>\n",
       "      <td>N</td>\n",
       "      <td>PDDQIGYYRRATRRIRGGDG</td>\n",
       "      <td>79</td>\n",
       "      <td>20</td>\n",
       "      <td>0.042728</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9462</th>\n",
       "      <td>N</td>\n",
       "      <td>PDDQIGYYRRATRRIRGGDGK</td>\n",
       "      <td>79</td>\n",
       "      <td>21</td>\n",
       "      <td>0.042728</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9861</th>\n",
       "      <td>N</td>\n",
       "      <td>PDDQIGYYRRATRRIRGGDGKM</td>\n",
       "      <td>79</td>\n",
       "      <td>22</td>\n",
       "      <td>0.044930</td>\n",
       "      <td>0.003597</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10259</th>\n",
       "      <td>N</td>\n",
       "      <td>PDDQIGYYRRATRRIRGGDGKMK</td>\n",
       "      <td>79</td>\n",
       "      <td>23</td>\n",
       "      <td>0.044930</td>\n",
       "      <td>0.003597</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10656</th>\n",
       "      <td>N</td>\n",
       "      <td>PDDQIGYYRRATRRIRGGDGKMKD</td>\n",
       "      <td>79</td>\n",
       "      <td>24</td>\n",
       "      <td>0.077581</td>\n",
       "      <td>0.006647</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11052</th>\n",
       "      <td>N</td>\n",
       "      <td>PDDQIGYYRRATRRIRGGDGKMKDL</td>\n",
       "      <td>79</td>\n",
       "      <td>25</td>\n",
       "      <td>0.077581</td>\n",
       "      <td>0.006647</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      protein                    epitope  start_pos  epi_len   entropy  \\\n",
       "4592        N                   PDDQIGYY         79        8  0.030754   \n",
       "5004        N                  PDDQIGYYR         79        9  0.030754   \n",
       "5415        N                 PDDQIGYYRR         79       10  0.031933   \n",
       "5825        N                PDDQIGYYRRA         79       11  0.039191   \n",
       "6234        N              PDDQIGYYRRATR         79       13  0.039191   \n",
       "6641        N             PDDQIGYYRRATRR         79       14  0.039191   \n",
       "7047        N            PDDQIGYYRRATRRI         79       15  0.039191   \n",
       "7452        N           PDDQIGYYRRATRRIR         79       16  0.040370   \n",
       "7856        N          PDDQIGYYRRATRRIRG         79       17  0.042728   \n",
       "8259        N         PDDQIGYYRRATRRIRGG         79       18  0.042728   \n",
       "8661        N        PDDQIGYYRRATRRIRGGD         79       19  0.042728   \n",
       "9062        N       PDDQIGYYRRATRRIRGGDG         79       20  0.042728   \n",
       "9462        N      PDDQIGYYRRATRRIRGGDGK         79       21  0.042728   \n",
       "9861        N     PDDQIGYYRRATRRIRGGDGKM         79       22  0.044930   \n",
       "10259       N    PDDQIGYYRRATRRIRGGDGKMK         79       23  0.044930   \n",
       "10656       N   PDDQIGYYRRATRRIRGGDGKMKD         79       24  0.077581   \n",
       "11052       N  PDDQIGYYRRATRRIRGGDGKMKDL         79       25  0.077581   \n",
       "\n",
       "       perc_mutated  glyco_probs  crosses_cleavage  South_African_mutation  \\\n",
       "4592       0.002581          0.0                 0                     0.0   \n",
       "5004       0.002581          0.0                 0                     0.0   \n",
       "5415       0.002659          0.0                 0                     0.0   \n",
       "5825       0.003206          0.0                 0                     0.0   \n",
       "6234       0.003206          0.0                 0                     0.0   \n",
       "6641       0.003206          0.0                 0                     0.0   \n",
       "7047       0.003206          0.0                 0                     0.0   \n",
       "7452       0.003284          0.0                 0                     0.0   \n",
       "7856       0.003441          0.0                 0                     0.0   \n",
       "8259       0.003441          0.0                 0                     0.0   \n",
       "8661       0.003441          0.0                 0                     0.0   \n",
       "9062       0.003441          0.0                 0                     0.0   \n",
       "9462       0.003441          0.0                 0                     0.0   \n",
       "9861       0.003597          0.0                 0                     0.0   \n",
       "10259      0.003597          0.0                 0                     0.0   \n",
       "10656      0.006647          0.0                 0                     0.0   \n",
       "11052      0.006647          0.0                 0                     0.0   \n",
       "\n",
       "       UK_mutation  Indian_mutation  Brazilian_mutation  \n",
       "4592           0.0              0.0                 1.0  \n",
       "5004           0.0              0.0                 1.0  \n",
       "5415           0.0              0.0                 1.0  \n",
       "5825           0.0              0.0                 1.0  \n",
       "6234           0.0              0.0                 1.0  \n",
       "6641           0.0              0.0                 1.0  \n",
       "7047           0.0              0.0                 1.0  \n",
       "7452           0.0              0.0                 1.0  \n",
       "7856           0.0              0.0                 1.0  \n",
       "8259           0.0              0.0                 1.0  \n",
       "8661           0.0              0.0                 1.0  \n",
       "9062           0.0              0.0                 1.0  \n",
       "9462           0.0              0.0                 1.0  \n",
       "9861           0.0              0.0                 1.0  \n",
       "10259          0.0              0.0                 1.0  \n",
       "10656          0.0              0.0                 1.0  \n",
       "11052          0.0              0.0                 1.0  "
      ]
     },
     "execution_count": 251,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df[np.logical_and(my_df.protein == 'N', my_df.start_pos == 80 - 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "id": "auburn-breath",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Code to save the data\n",
    "### my_df.to_csv('data/processed/AllEpitopeFeatures_with_South_African_and_UK_and_Indian_and_Brazilian_mutation_larger.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "consolidated-newfoundland",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "recent-ultimate",
   "metadata": {},
   "source": [
    "### Play around with the total data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "operational-buying",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_df = pd.read_csv('data/processed/AllEpitopeFeatures_with_South_African_and_UK_and_Indian_and_Brazilian_mutation_larger.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "id": "certified-cable",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0    159700\n",
       "1.0      4096\n",
       "Name: Brazilian_mutation, dtype: int64"
      ]
     },
     "execution_count": 257,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df.Brazilian_mutation.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "expressed-horse",
   "metadata": {},
   "source": [
    "### Set the mutation rate equal to 1 if the peptide is hit by (South_African, UK, Indian, Brazilian) mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "id": "unlimited-depth",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_df = pd.read_csv('data/processed/AllEpitopeFeatures_with_South_African_and_UK_and_Indian_and_Brazilian_mutation_larger.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "id": "incredible-collector",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>glyco_probs</th>\n",
       "      <th>crosses_cleavage</th>\n",
       "      <th>South_African_mutation</th>\n",
       "      <th>UK_mutation</th>\n",
       "      <th>Indian_mutation</th>\n",
       "      <th>Brazilian_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.035578</td>\n",
       "      <td>0.002893</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.042296</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.042296</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.041116</td>\n",
       "      <td>0.003363</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.038914</td>\n",
       "      <td>0.003206</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163791</th>\n",
       "      <td>S2</td>\n",
       "      <td>KGCCSCGSCCKFDEDDSEPVLKGVK</td>\n",
       "      <td>1244</td>\n",
       "      <td>25</td>\n",
       "      <td>0.116781</td>\n",
       "      <td>0.009853</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163792</th>\n",
       "      <td>S2</td>\n",
       "      <td>GCCSCGSCCKFDEDDSEPVLKGVKL</td>\n",
       "      <td>1245</td>\n",
       "      <td>25</td>\n",
       "      <td>0.113617</td>\n",
       "      <td>0.009618</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163793</th>\n",
       "      <td>S2</td>\n",
       "      <td>CCSCGSCCKFDEDDSEPVLKGVKLH</td>\n",
       "      <td>1246</td>\n",
       "      <td>25</td>\n",
       "      <td>0.113617</td>\n",
       "      <td>0.009618</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163794</th>\n",
       "      <td>S2</td>\n",
       "      <td>CSCGSCCKFDEDDSEPVLKGVKLHY</td>\n",
       "      <td>1247</td>\n",
       "      <td>25</td>\n",
       "      <td>0.099336</td>\n",
       "      <td>0.008367</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163795</th>\n",
       "      <td>S2</td>\n",
       "      <td>SCGSCCKFDEDDSEPVLKGVKLHYT</td>\n",
       "      <td>1248</td>\n",
       "      <td>25</td>\n",
       "      <td>0.098158</td>\n",
       "      <td>0.008289</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163796 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                    epitope  start_pos  epi_len   entropy  \\\n",
       "0            E                   MYSFVSEE          0        8  0.035578   \n",
       "1            E                   YSFVSEET          1        8  0.042296   \n",
       "2            E                   SFVSEETG          2        8  0.042296   \n",
       "3            E                   FVSEETGT          3        8  0.041116   \n",
       "4            E                   VSEETGTL          4        8  0.038914   \n",
       "...        ...                        ...        ...      ...       ...   \n",
       "163791      S2  KGCCSCGSCCKFDEDDSEPVLKGVK       1244       25  0.116781   \n",
       "163792      S2  GCCSCGSCCKFDEDDSEPVLKGVKL       1245       25  0.113617   \n",
       "163793      S2  CCSCGSCCKFDEDDSEPVLKGVKLH       1246       25  0.113617   \n",
       "163794      S2  CSCGSCCKFDEDDSEPVLKGVKLHY       1247       25  0.099336   \n",
       "163795      S2  SCGSCCKFDEDDSEPVLKGVKLHYT       1248       25  0.098158   \n",
       "\n",
       "        perc_mutated  glyco_probs  crosses_cleavage  South_African_mutation  \\\n",
       "0           0.002893          0.0                 0                     0.0   \n",
       "1           0.003441          0.0                 0                     0.0   \n",
       "2           0.003441          0.0                 0                     0.0   \n",
       "3           0.003363          0.0                 0                     0.0   \n",
       "4           0.003206          0.0                 0                     0.0   \n",
       "...              ...          ...               ...                     ...   \n",
       "163791      0.009853          0.0                 0                     0.0   \n",
       "163792      0.009618          0.0                 0                     0.0   \n",
       "163793      0.009618          0.0                 0                     0.0   \n",
       "163794      0.008367          0.0                 0                     0.0   \n",
       "163795      0.008289          0.0                 0                     0.0   \n",
       "\n",
       "        UK_mutation  Indian_mutation  Brazilian_mutation  \n",
       "0               0.0              0.0                 0.0  \n",
       "1               0.0              0.0                 0.0  \n",
       "2               0.0              0.0                 0.0  \n",
       "3               0.0              0.0                 0.0  \n",
       "4               0.0              0.0                 0.0  \n",
       "...             ...              ...                 ...  \n",
       "163791          0.0              0.0                 0.0  \n",
       "163792          0.0              0.0                 0.0  \n",
       "163793          0.0              0.0                 0.0  \n",
       "163794          0.0              0.0                 0.0  \n",
       "163795          0.0              0.0                 0.0  \n",
       "\n",
       "[163796 rows x 12 columns]"
      ]
     },
     "execution_count": 259,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "id": "neural-administration",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "40000\n",
      "50000\n",
      "60000\n",
      "70000\n",
      "80000\n",
      "90000\n",
      "100000\n",
      "110000\n",
      "120000\n",
      "130000\n",
      "140000\n",
      "150000\n",
      "160000\n"
     ]
    }
   ],
   "source": [
    "for ind in my_df.index:\n",
    "    if ind%10000==0:\n",
    "        print(ind)\n",
    "    if my_df.loc[ind,'UK_mutation'] == 1 or my_df.loc[ind,'South_African_mutation'] == 1 or my_df.loc[ind,'Indian_mutation'] == 1 or my_df.loc[ind,'Brazilian_mutation'] == 1:\n",
    "        my_df.loc[ind,'perc_mutated'] = 1.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "id": "industrial-cartridge",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df = my_df.drop(labels = ['South_African_mutation','UK_mutation', 'Indian_mutation', 'Brazilian_mutation'], axis='columns')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "id": "black-velvet",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>glyco_probs</th>\n",
       "      <th>crosses_cleavage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.035578</td>\n",
       "      <td>0.002893</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.042296</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.042296</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.041116</td>\n",
       "      <td>0.003363</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.038914</td>\n",
       "      <td>0.003206</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163791</th>\n",
       "      <td>S2</td>\n",
       "      <td>KGCCSCGSCCKFDEDDSEPVLKGVK</td>\n",
       "      <td>1244</td>\n",
       "      <td>25</td>\n",
       "      <td>0.116781</td>\n",
       "      <td>0.009853</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163792</th>\n",
       "      <td>S2</td>\n",
       "      <td>GCCSCGSCCKFDEDDSEPVLKGVKL</td>\n",
       "      <td>1245</td>\n",
       "      <td>25</td>\n",
       "      <td>0.113617</td>\n",
       "      <td>0.009618</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163793</th>\n",
       "      <td>S2</td>\n",
       "      <td>CCSCGSCCKFDEDDSEPVLKGVKLH</td>\n",
       "      <td>1246</td>\n",
       "      <td>25</td>\n",
       "      <td>0.113617</td>\n",
       "      <td>0.009618</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163794</th>\n",
       "      <td>S2</td>\n",
       "      <td>CSCGSCCKFDEDDSEPVLKGVKLHY</td>\n",
       "      <td>1247</td>\n",
       "      <td>25</td>\n",
       "      <td>0.099336</td>\n",
       "      <td>0.008367</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163795</th>\n",
       "      <td>S2</td>\n",
       "      <td>SCGSCCKFDEDDSEPVLKGVKLHYT</td>\n",
       "      <td>1248</td>\n",
       "      <td>25</td>\n",
       "      <td>0.098158</td>\n",
       "      <td>0.008289</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163796 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                    epitope  start_pos  epi_len   entropy  \\\n",
       "0            E                   MYSFVSEE          0        8  0.035578   \n",
       "1            E                   YSFVSEET          1        8  0.042296   \n",
       "2            E                   SFVSEETG          2        8  0.042296   \n",
       "3            E                   FVSEETGT          3        8  0.041116   \n",
       "4            E                   VSEETGTL          4        8  0.038914   \n",
       "...        ...                        ...        ...      ...       ...   \n",
       "163791      S2  KGCCSCGSCCKFDEDDSEPVLKGVK       1244       25  0.116781   \n",
       "163792      S2  GCCSCGSCCKFDEDDSEPVLKGVKL       1245       25  0.113617   \n",
       "163793      S2  CCSCGSCCKFDEDDSEPVLKGVKLH       1246       25  0.113617   \n",
       "163794      S2  CSCGSCCKFDEDDSEPVLKGVKLHY       1247       25  0.099336   \n",
       "163795      S2  SCGSCCKFDEDDSEPVLKGVKLHYT       1248       25  0.098158   \n",
       "\n",
       "        perc_mutated  glyco_probs  crosses_cleavage  \n",
       "0           0.002893          0.0                 0  \n",
       "1           0.003441          0.0                 0  \n",
       "2           0.003441          0.0                 0  \n",
       "3           0.003363          0.0                 0  \n",
       "4           0.003206          0.0                 0  \n",
       "...              ...          ...               ...  \n",
       "163791      0.009853          0.0                 0  \n",
       "163792      0.009618          0.0                 0  \n",
       "163793      0.009618          0.0                 0  \n",
       "163794      0.008367          0.0                 0  \n",
       "163795      0.008289          0.0                 0  \n",
       "\n",
       "[163796 rows x 8 columns]"
      ]
     },
     "execution_count": 262,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "id": "flexible-novel",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Code to save the new data!!!\n",
    "### new_df.to_csv('data/processed/AllEpitopeFeatures_accounting_for_all_major_mutations_mutrate_set_to_1.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "executed-marketplace",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>glyco_probs</th>\n",
       "      <th>crosses_cleavage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>142874</th>\n",
       "      <td>S1</td>\n",
       "      <td>LYRLFRKS</td>\n",
       "      <td>451</td>\n",
       "      <td>8</td>\n",
       "      <td>0.111980</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144140</th>\n",
       "      <td>S1</td>\n",
       "      <td>LYRLFRKSN</td>\n",
       "      <td>451</td>\n",
       "      <td>9</td>\n",
       "      <td>0.111980</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145405</th>\n",
       "      <td>S1</td>\n",
       "      <td>LYRLFRKSNL</td>\n",
       "      <td>451</td>\n",
       "      <td>10</td>\n",
       "      <td>0.111980</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146669</th>\n",
       "      <td>S1</td>\n",
       "      <td>LYRLFRKSNLK</td>\n",
       "      <td>451</td>\n",
       "      <td>11</td>\n",
       "      <td>0.113158</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147932</th>\n",
       "      <td>S1</td>\n",
       "      <td>LYRLFRKSNLKPF</td>\n",
       "      <td>451</td>\n",
       "      <td>13</td>\n",
       "      <td>0.114336</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149193</th>\n",
       "      <td>S1</td>\n",
       "      <td>LYRLFRKSNLKPFE</td>\n",
       "      <td>451</td>\n",
       "      <td>14</td>\n",
       "      <td>0.114336</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150453</th>\n",
       "      <td>S1</td>\n",
       "      <td>LYRLFRKSNLKPFER</td>\n",
       "      <td>451</td>\n",
       "      <td>15</td>\n",
       "      <td>0.114336</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151712</th>\n",
       "      <td>S1</td>\n",
       "      <td>LYRLFRKSNLKPFERD</td>\n",
       "      <td>451</td>\n",
       "      <td>16</td>\n",
       "      <td>0.114336</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152970</th>\n",
       "      <td>S1</td>\n",
       "      <td>LYRLFRKSNLKPFERDI</td>\n",
       "      <td>451</td>\n",
       "      <td>17</td>\n",
       "      <td>0.116537</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154227</th>\n",
       "      <td>S1</td>\n",
       "      <td>LYRLFRKSNLKPFERDIS</td>\n",
       "      <td>451</td>\n",
       "      <td>18</td>\n",
       "      <td>0.116537</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155483</th>\n",
       "      <td>S1</td>\n",
       "      <td>LYRLFRKSNLKPFERDIST</td>\n",
       "      <td>451</td>\n",
       "      <td>19</td>\n",
       "      <td>0.120624</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156738</th>\n",
       "      <td>S1</td>\n",
       "      <td>LYRLFRKSNLKPFERDISTE</td>\n",
       "      <td>451</td>\n",
       "      <td>20</td>\n",
       "      <td>0.121802</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157992</th>\n",
       "      <td>S1</td>\n",
       "      <td>LYRLFRKSNLKPFERDISTEI</td>\n",
       "      <td>451</td>\n",
       "      <td>21</td>\n",
       "      <td>0.121802</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159245</th>\n",
       "      <td>S1</td>\n",
       "      <td>LYRLFRKSNLKPFERDISTEIY</td>\n",
       "      <td>451</td>\n",
       "      <td>22</td>\n",
       "      <td>0.121802</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160497</th>\n",
       "      <td>S1</td>\n",
       "      <td>LYRLFRKSNLKPFERDISTEIYQ</td>\n",
       "      <td>451</td>\n",
       "      <td>23</td>\n",
       "      <td>0.121802</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161748</th>\n",
       "      <td>S1</td>\n",
       "      <td>LYRLFRKSNLKPFERDISTEIYQA</td>\n",
       "      <td>451</td>\n",
       "      <td>24</td>\n",
       "      <td>0.122980</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162998</th>\n",
       "      <td>S1</td>\n",
       "      <td>LYRLFRKSNLKPFERDISTEIYQAG</td>\n",
       "      <td>451</td>\n",
       "      <td>25</td>\n",
       "      <td>0.125337</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                    epitope  start_pos  epi_len   entropy  \\\n",
       "142874      S1                   LYRLFRKS        451        8  0.111980   \n",
       "144140      S1                  LYRLFRKSN        451        9  0.111980   \n",
       "145405      S1                 LYRLFRKSNL        451       10  0.111980   \n",
       "146669      S1                LYRLFRKSNLK        451       11  0.113158   \n",
       "147932      S1              LYRLFRKSNLKPF        451       13  0.114336   \n",
       "149193      S1             LYRLFRKSNLKPFE        451       14  0.114336   \n",
       "150453      S1            LYRLFRKSNLKPFER        451       15  0.114336   \n",
       "151712      S1           LYRLFRKSNLKPFERD        451       16  0.114336   \n",
       "152970      S1          LYRLFRKSNLKPFERDI        451       17  0.116537   \n",
       "154227      S1         LYRLFRKSNLKPFERDIS        451       18  0.116537   \n",
       "155483      S1        LYRLFRKSNLKPFERDIST        451       19  0.120624   \n",
       "156738      S1       LYRLFRKSNLKPFERDISTE        451       20  0.121802   \n",
       "157992      S1      LYRLFRKSNLKPFERDISTEI        451       21  0.121802   \n",
       "159245      S1     LYRLFRKSNLKPFERDISTEIY        451       22  0.121802   \n",
       "160497      S1    LYRLFRKSNLKPFERDISTEIYQ        451       23  0.121802   \n",
       "161748      S1   LYRLFRKSNLKPFERDISTEIYQA        451       24  0.122980   \n",
       "162998      S1  LYRLFRKSNLKPFERDISTEIYQAG        451       25  0.125337   \n",
       "\n",
       "        perc_mutated  glyco_probs  crosses_cleavage  \n",
       "142874           1.0          0.0                 0  \n",
       "144140           1.0          0.0                 0  \n",
       "145405           1.0          0.0                 0  \n",
       "146669           1.0          0.0                 0  \n",
       "147932           1.0          0.0                 0  \n",
       "149193           1.0          0.0                 0  \n",
       "150453           1.0          0.0                 0  \n",
       "151712           1.0          0.0                 0  \n",
       "152970           1.0          0.0                 0  \n",
       "154227           1.0          0.0                 0  \n",
       "155483           1.0          0.0                 0  \n",
       "156738           1.0          0.0                 0  \n",
       "157992           1.0          0.0                 0  \n",
       "159245           1.0          0.0                 0  \n",
       "160497           1.0          0.0                 0  \n",
       "161748           1.0          0.0                 0  \n",
       "162998           1.0          0.0                 0  "
      ]
     },
     "execution_count": 265,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df[np.logical_and(new_df.protein == 'S1', new_df.start_pos == 451)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "first-efficiency",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "nearby-newark",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "recreational-europe",
   "metadata": {},
   "source": [
    "## New code for setting the mutation rate for any peptide that has the UK or South African Mutation equal to 1 (can run this separately from the above)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "brief-drama",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_df = pd.read_csv('data/processed/AllEpitopeFeatures_with_South_African_and_UK_mutation_larger.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "exceptional-arlington",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>glyco_probs</th>\n",
       "      <th>crosses_cleavage</th>\n",
       "      <th>South_African_mutation</th>\n",
       "      <th>UK_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.035578</td>\n",
       "      <td>0.002893</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.042296</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.042296</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.041116</td>\n",
       "      <td>0.003363</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.038914</td>\n",
       "      <td>0.003206</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163791</th>\n",
       "      <td>S2</td>\n",
       "      <td>KGCCSCGSCCKFDEDDSEPVLKGVK</td>\n",
       "      <td>1244</td>\n",
       "      <td>25</td>\n",
       "      <td>0.116781</td>\n",
       "      <td>0.009853</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163792</th>\n",
       "      <td>S2</td>\n",
       "      <td>GCCSCGSCCKFDEDDSEPVLKGVKL</td>\n",
       "      <td>1245</td>\n",
       "      <td>25</td>\n",
       "      <td>0.113617</td>\n",
       "      <td>0.009618</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163793</th>\n",
       "      <td>S2</td>\n",
       "      <td>CCSCGSCCKFDEDDSEPVLKGVKLH</td>\n",
       "      <td>1246</td>\n",
       "      <td>25</td>\n",
       "      <td>0.113617</td>\n",
       "      <td>0.009618</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163794</th>\n",
       "      <td>S2</td>\n",
       "      <td>CSCGSCCKFDEDDSEPVLKGVKLHY</td>\n",
       "      <td>1247</td>\n",
       "      <td>25</td>\n",
       "      <td>0.099336</td>\n",
       "      <td>0.008367</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163795</th>\n",
       "      <td>S2</td>\n",
       "      <td>SCGSCCKFDEDDSEPVLKGVKLHYT</td>\n",
       "      <td>1248</td>\n",
       "      <td>25</td>\n",
       "      <td>0.098158</td>\n",
       "      <td>0.008289</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163796 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                    epitope  start_pos  epi_len   entropy  \\\n",
       "0            E                   MYSFVSEE          0        8  0.035578   \n",
       "1            E                   YSFVSEET          1        8  0.042296   \n",
       "2            E                   SFVSEETG          2        8  0.042296   \n",
       "3            E                   FVSEETGT          3        8  0.041116   \n",
       "4            E                   VSEETGTL          4        8  0.038914   \n",
       "...        ...                        ...        ...      ...       ...   \n",
       "163791      S2  KGCCSCGSCCKFDEDDSEPVLKGVK       1244       25  0.116781   \n",
       "163792      S2  GCCSCGSCCKFDEDDSEPVLKGVKL       1245       25  0.113617   \n",
       "163793      S2  CCSCGSCCKFDEDDSEPVLKGVKLH       1246       25  0.113617   \n",
       "163794      S2  CSCGSCCKFDEDDSEPVLKGVKLHY       1247       25  0.099336   \n",
       "163795      S2  SCGSCCKFDEDDSEPVLKGVKLHYT       1248       25  0.098158   \n",
       "\n",
       "        perc_mutated  glyco_probs  crosses_cleavage  South_African_mutation  \\\n",
       "0           0.002893          0.0                 0                     0.0   \n",
       "1           0.003441          0.0                 0                     0.0   \n",
       "2           0.003441          0.0                 0                     0.0   \n",
       "3           0.003363          0.0                 0                     0.0   \n",
       "4           0.003206          0.0                 0                     0.0   \n",
       "...              ...          ...               ...                     ...   \n",
       "163791      0.009853          0.0                 0                     0.0   \n",
       "163792      0.009618          0.0                 0                     0.0   \n",
       "163793      0.009618          0.0                 0                     0.0   \n",
       "163794      0.008367          0.0                 0                     0.0   \n",
       "163795      0.008289          0.0                 0                     0.0   \n",
       "\n",
       "        UK_mutation  \n",
       "0               0.0  \n",
       "1               0.0  \n",
       "2               0.0  \n",
       "3               0.0  \n",
       "4               0.0  \n",
       "...             ...  \n",
       "163791          0.0  \n",
       "163792          0.0  \n",
       "163793          0.0  \n",
       "163794          0.0  \n",
       "163795          0.0  \n",
       "\n",
       "[163796 rows x 10 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cross-treaty",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "40000\n",
      "50000\n",
      "60000\n",
      "70000\n",
      "80000\n",
      "90000\n",
      "100000\n",
      "110000\n",
      "120000\n",
      "130000\n",
      "140000\n",
      "150000\n",
      "160000\n"
     ]
    }
   ],
   "source": [
    "for ind in my_df.index:\n",
    "    if ind%10000==0:\n",
    "        print(ind)\n",
    "    if my_df.loc[ind,'UK_mutation'] == 1 or my_df.loc[ind,'South_African_mutation'] == 1:\n",
    "        my_df.loc[ind,'perc_mutated'] = 1.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fancy-criterion",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>glyco_probs</th>\n",
       "      <th>crosses_cleavage</th>\n",
       "      <th>South_African_mutation</th>\n",
       "      <th>UK_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.035578</td>\n",
       "      <td>0.002893</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.042296</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.042296</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.041116</td>\n",
       "      <td>0.003363</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.038914</td>\n",
       "      <td>0.003206</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163791</th>\n",
       "      <td>S2</td>\n",
       "      <td>KGCCSCGSCCKFDEDDSEPVLKGVK</td>\n",
       "      <td>1244</td>\n",
       "      <td>25</td>\n",
       "      <td>0.116781</td>\n",
       "      <td>0.009853</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163792</th>\n",
       "      <td>S2</td>\n",
       "      <td>GCCSCGSCCKFDEDDSEPVLKGVKL</td>\n",
       "      <td>1245</td>\n",
       "      <td>25</td>\n",
       "      <td>0.113617</td>\n",
       "      <td>0.009618</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163793</th>\n",
       "      <td>S2</td>\n",
       "      <td>CCSCGSCCKFDEDDSEPVLKGVKLH</td>\n",
       "      <td>1246</td>\n",
       "      <td>25</td>\n",
       "      <td>0.113617</td>\n",
       "      <td>0.009618</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163794</th>\n",
       "      <td>S2</td>\n",
       "      <td>CSCGSCCKFDEDDSEPVLKGVKLHY</td>\n",
       "      <td>1247</td>\n",
       "      <td>25</td>\n",
       "      <td>0.099336</td>\n",
       "      <td>0.008367</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163795</th>\n",
       "      <td>S2</td>\n",
       "      <td>SCGSCCKFDEDDSEPVLKGVKLHYT</td>\n",
       "      <td>1248</td>\n",
       "      <td>25</td>\n",
       "      <td>0.098158</td>\n",
       "      <td>0.008289</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163796 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                    epitope  start_pos  epi_len   entropy  \\\n",
       "0            E                   MYSFVSEE          0        8  0.035578   \n",
       "1            E                   YSFVSEET          1        8  0.042296   \n",
       "2            E                   SFVSEETG          2        8  0.042296   \n",
       "3            E                   FVSEETGT          3        8  0.041116   \n",
       "4            E                   VSEETGTL          4        8  0.038914   \n",
       "...        ...                        ...        ...      ...       ...   \n",
       "163791      S2  KGCCSCGSCCKFDEDDSEPVLKGVK       1244       25  0.116781   \n",
       "163792      S2  GCCSCGSCCKFDEDDSEPVLKGVKL       1245       25  0.113617   \n",
       "163793      S2  CCSCGSCCKFDEDDSEPVLKGVKLH       1246       25  0.113617   \n",
       "163794      S2  CSCGSCCKFDEDDSEPVLKGVKLHY       1247       25  0.099336   \n",
       "163795      S2  SCGSCCKFDEDDSEPVLKGVKLHYT       1248       25  0.098158   \n",
       "\n",
       "        perc_mutated  glyco_probs  crosses_cleavage  South_African_mutation  \\\n",
       "0           0.002893          0.0                 0                     0.0   \n",
       "1           0.003441          0.0                 0                     0.0   \n",
       "2           0.003441          0.0                 0                     0.0   \n",
       "3           0.003363          0.0                 0                     0.0   \n",
       "4           0.003206          0.0                 0                     0.0   \n",
       "...              ...          ...               ...                     ...   \n",
       "163791      0.009853          0.0                 0                     0.0   \n",
       "163792      0.009618          0.0                 0                     0.0   \n",
       "163793      0.009618          0.0                 0                     0.0   \n",
       "163794      0.008367          0.0                 0                     0.0   \n",
       "163795      0.008289          0.0                 0                     0.0   \n",
       "\n",
       "        UK_mutation  \n",
       "0               0.0  \n",
       "1               0.0  \n",
       "2               0.0  \n",
       "3               0.0  \n",
       "4               0.0  \n",
       "...             ...  \n",
       "163791          0.0  \n",
       "163792          0.0  \n",
       "163793          0.0  \n",
       "163794          0.0  \n",
       "163795          0.0  \n",
       "\n",
       "[163796 rows x 10 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "exciting-hughes",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>glyco_probs</th>\n",
       "      <th>crosses_cleavage</th>\n",
       "      <th>South_African_mutation</th>\n",
       "      <th>UK_mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>142433</th>\n",
       "      <td>S1</td>\n",
       "      <td>VSSQCVNL</td>\n",
       "      <td>10</td>\n",
       "      <td>8</td>\n",
       "      <td>0.208723</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142434</th>\n",
       "      <td>S1</td>\n",
       "      <td>SSQCVNLT</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>0.209743</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142435</th>\n",
       "      <td>S1</td>\n",
       "      <td>SQCVNLTT</td>\n",
       "      <td>12</td>\n",
       "      <td>8</td>\n",
       "      <td>0.198939</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142436</th>\n",
       "      <td>S1</td>\n",
       "      <td>QCVNLTTR</td>\n",
       "      <td>13</td>\n",
       "      <td>8</td>\n",
       "      <td>0.211701</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142437</th>\n",
       "      <td>S1</td>\n",
       "      <td>CVNLTTRT</td>\n",
       "      <td>14</td>\n",
       "      <td>8</td>\n",
       "      <td>0.228375</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163243</th>\n",
       "      <td>S2</td>\n",
       "      <td>MSLGAENSVAYSNNSIAIPTNFTIS</td>\n",
       "      <td>696</td>\n",
       "      <td>25</td>\n",
       "      <td>0.321672</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163244</th>\n",
       "      <td>S2</td>\n",
       "      <td>SLGAENSVAYSNNSIAIPTNFTISV</td>\n",
       "      <td>697</td>\n",
       "      <td>25</td>\n",
       "      <td>0.321672</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163245</th>\n",
       "      <td>S2</td>\n",
       "      <td>LGAENSVAYSNNSIAIPTNFTISVT</td>\n",
       "      <td>698</td>\n",
       "      <td>25</td>\n",
       "      <td>0.334331</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163246</th>\n",
       "      <td>S2</td>\n",
       "      <td>GAENSVAYSNNSIAIPTNFTISVTT</td>\n",
       "      <td>699</td>\n",
       "      <td>25</td>\n",
       "      <td>0.334331</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163247</th>\n",
       "      <td>S2</td>\n",
       "      <td>AENSVAYSNNSIAIPTNFTISVTTE</td>\n",
       "      <td>700</td>\n",
       "      <td>25</td>\n",
       "      <td>0.334331</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2216 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                    epitope  start_pos  epi_len   entropy  \\\n",
       "142433      S1                   VSSQCVNL         10        8  0.208723   \n",
       "142434      S1                   SSQCVNLT         11        8  0.209743   \n",
       "142435      S1                   SQCVNLTT         12        8  0.198939   \n",
       "142436      S1                   QCVNLTTR         13        8  0.211701   \n",
       "142437      S1                   CVNLTTRT         14        8  0.228375   \n",
       "...        ...                        ...        ...      ...       ...   \n",
       "163243      S2  MSLGAENSVAYSNNSIAIPTNFTIS        696       25  0.321672   \n",
       "163244      S2  SLGAENSVAYSNNSIAIPTNFTISV        697       25  0.321672   \n",
       "163245      S2  LGAENSVAYSNNSIAIPTNFTISVT        698       25  0.334331   \n",
       "163246      S2  GAENSVAYSNNSIAIPTNFTISVTT        699       25  0.334331   \n",
       "163247      S2  AENSVAYSNNSIAIPTNFTISVTTE        700       25  0.334331   \n",
       "\n",
       "        perc_mutated  glyco_probs  crosses_cleavage  South_African_mutation  \\\n",
       "142433           1.0          1.0                 0                     1.0   \n",
       "142434           1.0          1.0                 0                     1.0   \n",
       "142435           1.0          1.0                 0                     1.0   \n",
       "142436           1.0          1.0                 0                     1.0   \n",
       "142437           1.0          1.0                 0                     1.0   \n",
       "...              ...          ...               ...                     ...   \n",
       "163243           1.0          1.0                 0                     1.0   \n",
       "163244           1.0          1.0                 0                     1.0   \n",
       "163245           1.0          1.0                 0                     1.0   \n",
       "163246           1.0          1.0                 0                     1.0   \n",
       "163247           1.0          1.0                 0                     1.0   \n",
       "\n",
       "        UK_mutation  \n",
       "142433          0.0  \n",
       "142434          0.0  \n",
       "142435          0.0  \n",
       "142436          0.0  \n",
       "142437          0.0  \n",
       "...             ...  \n",
       "163243          1.0  \n",
       "163244          1.0  \n",
       "163245          1.0  \n",
       "163246          1.0  \n",
       "163247          1.0  \n",
       "\n",
       "[2216 rows x 10 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_df[my_df.UK_mutation == 1]\n",
    "my_df[my_df.South_African_mutation == 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "cardiovascular-determination",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df = my_df.drop(labels = ['South_African_mutation','UK_mutation'], axis='columns')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "binding-beast",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein</th>\n",
       "      <th>epitope</th>\n",
       "      <th>start_pos</th>\n",
       "      <th>epi_len</th>\n",
       "      <th>entropy</th>\n",
       "      <th>perc_mutated</th>\n",
       "      <th>glyco_probs</th>\n",
       "      <th>crosses_cleavage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>MYSFVSEE</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>0.035578</td>\n",
       "      <td>0.002893</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>E</td>\n",
       "      <td>YSFVSEET</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>0.042296</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>E</td>\n",
       "      <td>SFVSEETG</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.042296</td>\n",
       "      <td>0.003441</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>E</td>\n",
       "      <td>FVSEETGT</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0.041116</td>\n",
       "      <td>0.003363</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>E</td>\n",
       "      <td>VSEETGTL</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.038914</td>\n",
       "      <td>0.003206</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163791</th>\n",
       "      <td>S2</td>\n",
       "      <td>KGCCSCGSCCKFDEDDSEPVLKGVK</td>\n",
       "      <td>1244</td>\n",
       "      <td>25</td>\n",
       "      <td>0.116781</td>\n",
       "      <td>0.009853</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163792</th>\n",
       "      <td>S2</td>\n",
       "      <td>GCCSCGSCCKFDEDDSEPVLKGVKL</td>\n",
       "      <td>1245</td>\n",
       "      <td>25</td>\n",
       "      <td>0.113617</td>\n",
       "      <td>0.009618</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163793</th>\n",
       "      <td>S2</td>\n",
       "      <td>CCSCGSCCKFDEDDSEPVLKGVKLH</td>\n",
       "      <td>1246</td>\n",
       "      <td>25</td>\n",
       "      <td>0.113617</td>\n",
       "      <td>0.009618</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163794</th>\n",
       "      <td>S2</td>\n",
       "      <td>CSCGSCCKFDEDDSEPVLKGVKLHY</td>\n",
       "      <td>1247</td>\n",
       "      <td>25</td>\n",
       "      <td>0.099336</td>\n",
       "      <td>0.008367</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163795</th>\n",
       "      <td>S2</td>\n",
       "      <td>SCGSCCKFDEDDSEPVLKGVKLHYT</td>\n",
       "      <td>1248</td>\n",
       "      <td>25</td>\n",
       "      <td>0.098158</td>\n",
       "      <td>0.008289</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>163796 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       protein                    epitope  start_pos  epi_len   entropy  \\\n",
       "0            E                   MYSFVSEE          0        8  0.035578   \n",
       "1            E                   YSFVSEET          1        8  0.042296   \n",
       "2            E                   SFVSEETG          2        8  0.042296   \n",
       "3            E                   FVSEETGT          3        8  0.041116   \n",
       "4            E                   VSEETGTL          4        8  0.038914   \n",
       "...        ...                        ...        ...      ...       ...   \n",
       "163791      S2  KGCCSCGSCCKFDEDDSEPVLKGVK       1244       25  0.116781   \n",
       "163792      S2  GCCSCGSCCKFDEDDSEPVLKGVKL       1245       25  0.113617   \n",
       "163793      S2  CCSCGSCCKFDEDDSEPVLKGVKLH       1246       25  0.113617   \n",
       "163794      S2  CSCGSCCKFDEDDSEPVLKGVKLHY       1247       25  0.099336   \n",
       "163795      S2  SCGSCCKFDEDDSEPVLKGVKLHYT       1248       25  0.098158   \n",
       "\n",
       "        perc_mutated  glyco_probs  crosses_cleavage  \n",
       "0           0.002893          0.0                 0  \n",
       "1           0.003441          0.0                 0  \n",
       "2           0.003441          0.0                 0  \n",
       "3           0.003363          0.0                 0  \n",
       "4           0.003206          0.0                 0  \n",
       "...              ...          ...               ...  \n",
       "163791      0.009853          0.0                 0  \n",
       "163792      0.009618          0.0                 0  \n",
       "163793      0.009618          0.0                 0  \n",
       "163794      0.008367          0.0                 0  \n",
       "163795      0.008289          0.0                 0  \n",
       "\n",
       "[163796 rows x 8 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "graduate-prefix",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df.to_csv('data/processed/AllEpitopeFeatures_accounting_for_UK_and_SA_mutations.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "authentic-handle",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
